src/main/java/net/fabricmc/loom/providers/LaunchProvider.java

```diff
@@ -0,0 +1,275 @@
+
+package net.fabricmc.loom.providers;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+import com.google.common.io.ByteStreams;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import org.apache.commons.io.FileUtils;
+import org.gradle.api.Project;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.ClassNode;
+import org.objectweb.asm.tree.InsnList;
+import org.objectweb.asm.tree.InsnNode;
+import org.objectweb.asm.tree.LdcInsnNode;
+import org.objectweb.asm.tree.MethodInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.VarInsnNode;
+import org.zeroturnaround.zip.ByteSource;
+import org.zeroturnaround.zip.ZipEntrySource;
+import org.zeroturnaround.zip.ZipUtil;
+
+import net.fabricmc.launch.Launch;
+import net.fabricmc.loom.LoomGradleExtension;
+import net.fabricmc.loom.util.Constants;
+import net.fabricmc.loom.util.DependencyProvider;
+
+public class LaunchProvider extends DependencyProvider {
+	@Override
+	public void provide(DependencyInfo dependency, Project project, LoomGradleExtension extension, Consumer<Runnable> postPopulationScheduler) {
+		if (extension.getLoaderLaunchMethod().equals("launchwrapper")) {
+			return;
+		}
+
+		postPopulationScheduler.accept(() -> {
+			File jarFile = new File(extension.getRootProjectPersistentCache(), "LaunchClasses.jar");
+
+			if (jarFile.exists()) {
+				//TODO possibly check the jar to see if it does need changing? some sort of hash?
+				boolean success = jarFile.delete();
+				if (!success) throw new RuntimeException("Failed to delete launchclasses.jar, is the game currently running?");
+			}
+
+			final LaunchDetails clientLaunchDetails = new LaunchDetails(getMainClass("client", extension));
+			final LaunchDetails serverLaunchDetails = new LaunchDetails(getMainClass("server", extension));
+
+			clientLaunchDetails
+					.property("fabric.development", "true")
+					.property("java.library.path", extension.getNativesDirectory().getAbsolutePath())
+					.property("org.lwjgl.librarypath", extension.getNativesDirectory().getAbsolutePath())
+					.argument("--assetIndex")
+					.argument(extension.getMinecraftProvider().versionInfo.assetIndex.getFabricId(extension.getMinecraftProvider().minecraftVersion))
+					.argument("--assetsDir")
+					.argument(new File(extension.getUserCache(), "assets").getAbsolutePath());
+
+			serverLaunchDetails
+					.property("fabric.development", "true");
+
+			final List<ZipEntrySource> entries = new ArrayList<>();
+			entries.add(generateZipEntry("net/fabricmc/launch/LaunchClient", clientLaunchDetails));
```

Comment:
`new ZipEntrySource[]{generateZipEntry...., generateZipEntry...}` instead of this list ordeal